### PR TITLE
#87 - Added S3Storage with save() method implemented

### PR DESCRIPTION
**Pull request by olegmoz**
Part of issue #87 
Added S3Storage class skeleton with `save()` method implementation.
Uploading is done using multipart upload, because content length has to be specified when using  S# put request and it cannot be calculated without loading all content in memory with existing `Storage` interface.

### CODE REVIEW

**Comment by 0crat** [comment]
This pull request [#89](https://github.com/artipie/asto/issues/89) is assigned to @hdouss[/z](https://www.0crat.com/u/hdouss), here is [why](https://www.0crat.com/footprint/CT2E6TK9B/6dc110aa-d8db-440f-a659-0b96e83923de); the budget is 15 minutes, see [§4](http://www.zerocracy.com/policy.html#4); please, read [§27](http://www.zerocracy.com/policy.html#27) and when you decide to accept the changes, inform @g4s8[/z](https://www.0crat.com/u/g4s8) (the architect) right in this ticket; if you decide that this PR should not be accepted ever, also inform the architect; this [blog post](http://www.yegor256.com/2015/02/09/serious-code-reviewer.html) will help you understand what is expected from a code reviewer; there will be a monetary reward for this job

File: src/main/java/com/artipie/asto/s3/MultipartUpload.java [hunk 1]
```diff
@@ -0,0 +1,177 @@
...
+     * @param key S3 object key.
+     * @param id ID of this upload.
+     * @checkstyle ParameterNumberCheck (2 line)
+     */
+    MultipartUpload(
+        final S3AsyncClient client,
+        final String bucket,
+        final Key key,
+        final String id) {
+        this.client = client;
```
Comment (HDouss) [inline src/main/java/com/artipie/asto/s3/MultipartUpload.java]: @olegmoz I think we should have a puzzle here to implement an upload data holder class. The puzzle should mention to get rid of `@checkstyle ParameterNumberCheck (2 line)`

File: src/main/java/com/artipie/asto/s3/S3Storage.java [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+     * @param bucket Bucket name.
+     */
+    public S3Storage(final S3AsyncClient client, final String bucket) {
+        this.client = client;
+        this.bucket = bucket;
+    }
+
+    @Override
+    public CompletableFuture<Boolean> exists(final Key key) {
+        throw new UnsupportedOperationException();
```
Comment (HDouss) [inline src/main/java/com/artipie/asto/s3/S3Storage.java]: @olegmoz Should we add a puzzle to support this in the future?

File: src/main/java/com/artipie/asto/s3/S3Storage.java [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+    }
+
+    @Override
+    public CompletableFuture<Publisher<ByteBuffer>> value(final Key key) {
+        throw new UnsupportedOperationException();
+    }
+
+    @Override
+    public CompletableFuture<Transaction> transaction(final List<Key> keys) {
+        throw new UnsupportedOperationException();
```
Comment (HDouss) [inline src/main/java/com/artipie/asto/s3/S3Storage.java]: @olegmoz Same for all unsupported operations: I suggest to add a puzzle to support what can be supported in the future.

**Comment by olegmoz** [comment]
@HDouss thanks for youe input! Added puzzle for `MultipartUpload` class refactoring, but do not want to add puzzles for each of S3Storage methods. Issue #87 won't be closed until all of them are implemented and this approach (many PRs per issue) is normal practice in this project.

**Comment by HDouss** [comment]
@olegmoz Ok for me if @g4s8 approves that. It is just that puzzles makes issues creation automatic and more suitable for micro tasking approach.

**Comment by HDouss** [comment]
@olegmoz branch is out of date

**Comment by olegmoz** [comment]
@HDouss merged with master, there were no conflicts though
@g4s8 please review

**Comment by 0crat** [comment]
Job was finished in 5 hours, bonus for fast delivery is possible (see [§36](http://www.zerocracy.com/policy.html#36))

**Comment by 0crat** [comment]
Job audit: `ARC` didn't merge or close PR, performer @hdouss[/z](https://www.0crat.com/u/hdouss)[/z](https://www.0crat.com/u/hdouss) didn't make CR comments

**Pull request merged**